src/Agent.Worker/Release/Artifacts/BuildArtifact.cs

```diff
@@ -0,0 +1,242 @@
+                                       parallelDownloadLimit,
+                                       CultureInfo.InvariantCulture)
+                };
+            }
+            else
+            {
+                throw new InvalidOperationException(StringUtil.Loc("RMArtifactDetailsIncomplete"));
+            }
+        }
+
+        private bool Match(ServerBuildArtifact buildArtifact, ArtifactDefinition artifactDefinition)
+        {
+            // If this is older , then dont force the name checks. 
+            if (string.Equals(artifactDefinition.Name, AllArtifacts, StringComparison.OrdinalIgnoreCase))
+            {
+                return true;
+            }
+
+            if (string.Equals(artifactDefinition.Name, buildArtifact.Name, StringComparison.OrdinalIgnoreCase))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private async Task DownloadArtifact(
+            IExecutionContext executionContext,
+            ServerBuildArtifact buildArtifact,
+            ArtifactDefinition artifactDefinition,
+            string localFolderPath,
+            BuildHttpClient buildClient,
+            XamlBuildHttpClient xamlBuildClient,
+            DefinitionType definitionType,
+            int buildId)
+        {
+            var downloadFolderPath = Path.Combine(localFolderPath, buildArtifact.Name);
+            var buildArtifactDetails = artifactDefinition.Details as BuildArtifactDetails;
+            if ((buildArtifact.Resource.Type == null && buildArtifact.Id == 0) // bug on build API Bug 378900
+                || string.Equals(buildArtifact.Resource.Type, WellKnownArtifactResourceTypes.FilePath, StringComparison.OrdinalIgnoreCase))
+            {
+                executionContext.Output("Artifact Type: FileShare");
+                string fileShare;
+                if (buildArtifact.Id == 0)
+                {
+                    fileShare = new Uri(buildArtifact.Resource.DownloadUrl).LocalPath;
+                }
+                else
+                {
+                    fileShare = new Uri(Path.Combine(buildArtifact.Resource.DownloadUrl, buildArtifact.Name)).LocalPath;
+                    if (!Directory.Exists(fileShare))
+                    {
+                        // download path does not exist, log and fall back
+                        executionContext.Output(StringUtil.Loc("RMArtifactNameDirectoryNotFound", fileShare, buildArtifact.Resource.DownloadUrl));
+                        fileShare = new Uri(buildArtifact.Resource.DownloadUrl).LocalPath;
+                    }
+                }
+
+                if (!Directory.Exists(fileShare))
+                {
+                    // download path does not exist, raise exception
+                    throw new ArtifactDownloadException(StringUtil.Loc("RMArtifactDirectoryNotFoundError", fileShare));
+                }
+
+                var fileShareArtifact = new FileShareArtifact();
```

Comment:
FileShareArtifact is an AgentService. Is there a reason not to use GetService<T> here?